### ISSUE TITLE
Values converge to 1.0 then RuntimeWarning

### ISSUE THREAD

**Issue by alexjc** [duguyue100/conceptors#1]
I tried plugging in the CIFAR10 data normalized by reproducing your `util.load_CIFAR10` function (which is not in the repository), and after one step of the inner training loop `x` converges to 1.0 which throws a warning:

```
conceptors/conceptors/net.py:325: RuntimeWarning: divide by zero encountered in arctanh
  self.W_targets=np.arctanh(self.all_train_args)-numpy.matlib.repmat(self.W_bias, 1, self.num_pattern*self.learn_length);
```

At this stage, `self.all_train_args` has minimum value 1.0 and maximum value 1.0.  Could you either commit your modified utils module, or let me know how to fix this?

Thanks!

**Comment by alexjc** [comment]
I think I found your code ;-)
https://github.com/duguyue100/cs231n-practice/blob/master/cs231nlib/utils.py

Trying it now...

**Comment by alexjc** [comment]
I added the exact copy of `load_CIFAR10` that you used before, the example launches fine but the `RuntimeWarning` above persists for the exact same reasons.

**Comment by duguyue100** [comment]
Hi,

Thanks for running my code.. I didn't actually expect anyone would run my messy code right now..
Yes, my `load_CIFAR10` is simply adapted from cs231's utility functions.

You are probably running `classify_test.py`. This is not a successful run. I didn't remove it since I thought I should at least leave a record for myself.

The correct one you are looking for is `classify_exp`. However this one does not take CIFAR10's raw image to classify, it takes features from feature extractor (auto-encoder, convnet or whatever). You probably want to do that for your self. Data's structure is like

```
label1 feature1
label2 feature2
...
```

they are organised in to a full matrix.

And for convenience, I pre-ordered the training dataset since I need to calculate conceptors for each class.

And in order to log and control the experiment with a .json file, you need to use `sacred` as an external dependencies.

Right now it is still a little bit messy, you may run into some dirts while you are digging. Have fun.

Regards,
Hu Yuhuang.

**Comment by duguyue100** [comment]
Aha.. I thought I didn't answer your question, yes, I'm aware of this warning.
`W_targets` is one of the variable that existed in original theory, I actually found it's useless in classification, but it could be useful for other things (like re-generate signal without source), so I kept avoiding it while doing my current research.

**Comment by alexjc** [comment]
Thanks! Going to see what I can come up with...

**Comment by alexjc** [comment]
The exact format of the array, is it a numpy.array of (N, M+1) where N is the number of samples and M is the number of features? Then one additional item at index `0` for the label?

**Comment by duguyue100** [comment]
yeah, exactly.

and in case you want to know how I manipulate with training set:
https://github.com/duguyue100/telauges/blob/master/scripts/cifar10_convnet_feature.py#L185-L198

Please don't use the code entirely, it's not stable, you need to configure a lot for making it works. It supposed to be a tutorial project for a neural network class.

**Comment by alexjc** [comment]
OK, it's basically working, now I just have to tweak the quality of the features.

Based on the paper:
- What activations did you use for the auto-encoder, and how long did you train it?
- For the convolution network, was that the architecture you used to get 98% on CIFAR10?

Thanks!

**Comment by duguyue100** [comment]
- for auto-encoder, it's just an one-hidden layer structure, and I used 200 hidden nodes and take the features.
- for convnet, I also used a simple one, two convolution layer and one fully connected layers and one softmax layer (trained supervised first). first convolution layer as 50 filters of (7,7), and second is 20 filters of (4,4). fully connect layer has 200 nodes and finally softmax has 10 nodes, I used fully connect layer's output as feature.

all structures are trained with 100 epochs, and each batch size is 200.

Actually feature extraction does some part of job, but a shallow one like this should work

**Comment by alexjc** [comment]
Thanks!

Is the rest of the code in the repository what you used in the paper? I'm currently at 38% accuracy after training an auto-encoder for 15 epochs...

**Comment by duguyue100** [comment]
The rest code are either for another kind of experiments or not using at all for the paper.

It's strange for me that get low accuracy (especially this low). did you pre-order the training dataset? since I want the conceptor as simpler as possible, this is a pre-request for the code.
another way is to train more epochs for your auto-encoder.

**Comment by duguyue100** [comment]
if it's still not working, then I can send you my data for you to test.

**Comment by alexjc** [comment]
Yes, I ordered the training data but not the test data.  Now reaching 40%.  I just realized I have color values in my data, where you were using greyscale...

I use an auto-encoder that's got a `Tanh` activation, I tried 512 units trained for 100 iterations with batch size 200 and learning rate 0.002.  I will try training it a bit longer.

My best run was with 20 Conceptor neurons and 256 features from the autoencoder.

```
INFO - Classification Task - Running command 'classify_experiment'
INFO - Classification Task - Started
[MESSAGE] Data is prepared.
[MESSAGE] Trail 1 --- Number of Neuron: 20
[MESSAGE] Train data driven
[MESSAGE] Conceptors Computed
[MESSAGE] Best Positive Aperture: 4.98, Best Negative Aperture: 4.70
[MESSAGE] Best conceptors computed
[MESSAGE] 1 class evidence is calculated
[MESSAGE] 2 class evidence is calculated
[MESSAGE] 3 class evidence is calculated
[MESSAGE] 4 class evidence is calculated
[MESSAGE] 5 class evidence is calculated
[MESSAGE] 6 class evidence is calculated
[MESSAGE] 7 class evidence is calculated
[MESSAGE] 8 class evidence is calculated
[MESSAGE] 9 class evidence is calculated
[MESSAGE] 10 class evidence is calculated
[MESSAGE] Accuracy 40.09 %
[MESSAGE] Total for 5.37m
```
- How did you train the convolution network, on its own input data auto-encoder style?
- What aperture values should I bet getting?

Thanks for all your help!  I think it's awesome you're making your code open.

**Comment by duguyue100** [comment]
- for convolution network, I simply trained it in supervised learning and then obtained the output from the fully connected layer before softmax layer.
- the aperture is an old name in the original theory for regularization rate. Originally we thought this value is important, so we used a simple idea to find it. but then during the training, when I saw the data and running result, I realised that it's not that important.

What you got here is interesting. I guess one reason is that my configuration and yours has some slightly difference in some way. I will run more tests tomorrow to give you an answer. The bottom line is, I also want to know that I'm not working this thing out by coincidence

**Comment by duguyue100** [comment]
Sorry for the delay, but I re-run my program just now, and I didn't find any problems. Here is a little thing I changed in my program: previously, one version of my code use ordered test file, that means I even pre-order the testing. This seems like a useless move afterward. Therefore, this time, I random shuffle the rows of data either I'm using ordered or random order data.

I just tested it, it doesn't affect the accuracy (of course it's not, what I'm thinking).

Then, the remain problem maybe your data. If you can write me an email, I will send you my AE and ConvNet feature for you to test on. My email is duguyue100@gmail.com

**Comment by alexjc** [comment]
Thanks!

After converting to greyscale yesterday, performance dropped.  I looked at the input data and it seems visually correct.  I've been training a convolution network instead of an auto-encoder, but it's quite slow on this machine... :-)

I'll make my entire scripts available soon, that would be easiest, so there's something public we can work on.
